{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sbn\n",
    "from Bio import SeqIO\n",
    "import SearchTools\n",
    "\n",
    "%matplotlib inline\n",
    "sbn.set(font_scale=1.5)\n",
    "sbn.set_style('white')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Import\n",
    "\n",
    "Sequence data is read in from the Los Alamos Database that is stored in the included `.tar.gz` file. These alignments were produced by the LANL tools and are aranged such that the HXB2 sequence is the last in the file. The alignment information is used to determine the HXB2 positions of each sequence. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing hiv-2500-3000.fixed.fst\n",
      "Importing hiv-1100-1950.fixed.fst\n",
      "Importing hiv-8300-8900.fixed.fst\n",
      "Importing hiv-7100-7500.fixed.fst\n",
      "Importing hiv-5200-5600.fixed.fst\n",
      "Importing hiv-1-700.fixed.fst\n",
      "Importing hiv-8500-8800.fixed.fst\n",
      "Importing hiv-6200-6900.fixed.fst\n",
      "Importing hiv-700-1150.fixed.fst\n",
      "Importing hiv-2000-2500.fixed.fst\n",
      "Importing hiv-5700-6100.fixed.fst\n",
      "Importing hiv-9086-9717.fixed.fst\n",
      "Importing hiv-4100-4900.fixed.fst\n",
      "Importing hiv-3000-3500.fixed.fst\n",
      "Importing hiv-7500-7900.fixed.fst\n",
      "Importing hiv-3200-3500.fixed.fst\n",
      "Total sequences imported 390290\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>SeqR</th>\n",
       "      <th>sStart</th>\n",
       "      <th>sStop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>B.US.1997.ARES2.AB078005</td>\n",
       "      <td>(A, C, A, T, A, A, T, T, G, -, -, -, -, -, G, ...</td>\n",
       "      <td>2500</td>\n",
       "      <td>3000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>B.US.1985.Ba_L.AB221005</td>\n",
       "      <td>(A, C, A, T, A, A, T, T, G, -, -, -, -, -, G, ...</td>\n",
       "      <td>2500</td>\n",
       "      <td>3000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B.US.1985.Ba_L.AB253432</td>\n",
       "      <td>(A, C, A, T, A, A, T, T, G, -, -, -, -, -, G, ...</td>\n",
       "      <td>2500</td>\n",
       "      <td>3000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>B.US.1991.US2.AB485638</td>\n",
       "      <td>(A, C, A, T, A, A, T, T, G, -, -, -, -, -, G, ...</td>\n",
       "      <td>2500</td>\n",
       "      <td>3000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B.US.1991.US2.AB485639</td>\n",
       "      <td>(A, C, A, T, A, A, T, T, G, -, -, -, -, -, G, ...</td>\n",
       "      <td>2500</td>\n",
       "      <td>3000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Name  \\\n",
       "0  B.US.1997.ARES2.AB078005   \n",
       "1   B.US.1985.Ba_L.AB221005   \n",
       "2   B.US.1985.Ba_L.AB253432   \n",
       "3    B.US.1991.US2.AB485638   \n",
       "4    B.US.1991.US2.AB485639   \n",
       "\n",
       "                                                SeqR  sStart  sStop  \n",
       "0  (A, C, A, T, A, A, T, T, G, -, -, -, -, -, G, ...    2500   3000  \n",
       "1  (A, C, A, T, A, A, T, T, G, -, -, -, -, -, G, ...    2500   3000  \n",
       "2  (A, C, A, T, A, A, T, T, G, -, -, -, -, -, G, ...    2500   3000  \n",
       "3  (A, C, A, T, A, A, T, T, G, -, -, -, -, -, G, ...    2500   3000  \n",
       "4  (A, C, A, T, A, A, T, T, G, -, -, -, -, -, G, ...    2500   3000  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import tarfile\n",
    "from collections import Counter\n",
    "from copy import deepcopy\n",
    "from itertools import compress\n",
    "from collections import deque\n",
    "\n",
    "def get_start_stop(seqR, hxb2_pos):\n",
    "    \"\"\"Determine the HXB2 start-stop of the sequence using affine-gaps.\"\"\"\n",
    "    \n",
    "    it = zip(str(seqR.seq), hxb2_pos)\n",
    "    poses = compress(it, (x[0] != '-' for x in it))\n",
    "    start = poses.next()[1]\n",
    "    stop = deque(poses, maxlen=1)[0][1]\n",
    "    return start, stop\n",
    "\n",
    "\n",
    "seq_locs = []\n",
    "\n",
    "with tarfile.open('data/LANLdata.tar.gz', mode='r:gz') as tr:\n",
    "    for f in tr:\n",
    "        \n",
    "        parts = f.name.split('/')[-1].split('.')[0].split('-')\n",
    "        offset = int(parts[1])\n",
    "        print 'Importing', f.name\n",
    "        \n",
    "        seqs = list(SeqIO.parse(tr.extractfile(f), 'fasta'))\n",
    "        hxb2 = seqs[-1]\n",
    "\n",
    "        hxb2_pos = []\n",
    "        pos = offset-1\n",
    "        for hx_l in str(hxb2.seq):\n",
    "            pos += hx_l != '-'\n",
    "            hxb2_pos.append(pos)\n",
    "\n",
    "        for seqR in seqs[:-1]:\n",
    "            if len(seqR.seq.ungap('-')) > 50:\n",
    "                start, stop = get_start_stop(seqR, hxb2_pos)\n",
    "                seq_locs.append({'sStart': start,\n",
    "                                 'sStop': stop,\n",
    "                                 'SeqR': deepcopy(seqR),\n",
    "                                 'Name': seqR.id})\n",
    "\n",
    "            \n",
    "seq_df = pd.DataFrame(seq_locs)\n",
    "print 'Total sequences imported', len(seq_locs)\n",
    "seq_df.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Citation</th>\n",
       "      <th>Name</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>gRNA</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>26607397</td>\n",
       "      <td>LTR-1</td>\n",
       "      <td>28</td>\n",
       "      <td>47</td>\n",
       "      <td>GACAAGATATCCTTGATCTG.NGG</td>\n",
       "      <td>LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26607397</td>\n",
       "      <td>LTR-2</td>\n",
       "      <td>78</td>\n",
       "      <td>97</td>\n",
       "      <td>GATTGACAGAACTACACACC.NGG</td>\n",
       "      <td>LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>26607397</td>\n",
       "      <td>LTR-3</td>\n",
       "      <td>108</td>\n",
       "      <td>127</td>\n",
       "      <td>GTCAGATATCCACTGACCTT.NGG</td>\n",
       "      <td>LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>26607397</td>\n",
       "      <td>LTR-4</td>\n",
       "      <td>343</td>\n",
       "      <td>362</td>\n",
       "      <td>GCTACAAGGGACTTTCCGCT.NGG</td>\n",
       "      <td>LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>26607397</td>\n",
       "      <td>LTR-5</td>\n",
       "      <td>379</td>\n",
       "      <td>398</td>\n",
       "      <td>GGCGTGGCCTGGGCGGGACT.NGG</td>\n",
       "      <td>LTR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Citation   Name  Start  Stop                      gRNA Region\n",
       "0  26607397  LTR-1     28    47  GACAAGATATCCTTGATCTG.NGG    LTR\n",
       "1  26607397  LTR-2     78    97  GATTGACAGAACTACACACC.NGG    LTR\n",
       "2  26607397  LTR-3    108   127  GTCAGATATCCACTGACCTT.NGG    LTR\n",
       "3  26607397  LTR-4    343   362  GCTACAAGGGACTTTCCGCT.NGG    LTR\n",
       "4  26607397  LTR-5    379   398  GGCGTGGCCTGGGCGGGACT.NGG    LTR"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ex_df = pd.read_excel('data/gRNAList.xlsx')\n",
    "ex_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sequence Processing\n",
    "\n",
    "Now that the gRNAs and sequences have been loaded we'll compare them. Each gRNA will be parsed to extract the protospacer region. For each gRNA we extract all sequences that overlap the target region. Then, the MIT penalty matrix is applied exauhstively across the entire sequence to determine the ideal binding location."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from Bio.Seq import reverse_complement\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_grna(gRNA):\n",
    "    \"\"\"Extract and normalize the length of the protospacer\"\"\"\n",
    "    \n",
    "    parts = gRNA.split('.')\n",
    "    if len(parts[1]) == 3:\n",
    "        direc = 'Forward'\n",
    "        proto = parts[0][-20:]\n",
    "    else:\n",
    "        direc = 'Reverse'\n",
    "        proto = parts[1][:20]\n",
    "        \n",
    "    if len(proto) < 20:\n",
    "        proto = 'N'*(20-len(proto)) + proto\n",
    "        \n",
    "    return direc, proto\n",
    "\n",
    "\n",
    "def check_whole_seq(seq, gRNA):\n",
    "    \"\"\"Apply the MIT penalty matrix across every postion on the sequence.\"\"\"\n",
    "    \n",
    "    direc, query = parse_grna(gRNA)\n",
    "    \n",
    "    # Normalize the orientation and ungap the sequence\n",
    "    if direc == 'Reverse':\n",
    "        nseq = seq.ungap('-')[::-1]\n",
    "        query = query[::-1]\n",
    "    else:\n",
    "        nseq = seq.ungap('-')\n",
    "    \n",
    "    # Check every 20-mer in the sequence.\n",
    "    scores = []\n",
    "    for start in range(len(nseq)-20):\n",
    "        scores.append(SearchTools.gRNA_score_hit(query, nseq[start:start+20]))\n",
    "\n",
    "    # Find the best score\n",
    "    scores = pd.Series(scores)\n",
    "    bscore, bloc = scores.max(), scores.idxmax()\n",
    "    \n",
    "    # Prep outputs\n",
    "    oseq = str(nseq[bloc:bloc+23])\n",
    "    has_pam = oseq.endswith('GG')\n",
    "    if direc == 'Reverse':\n",
    "        oseq = oseq.encode('ascii')[::-1]\n",
    "        has_pam = oseq.startswith('CC')\n",
    "    \n",
    "    return oseq, bscore, direc, has_pam\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50000 gEnv1 7216\n",
      "100000 gGag1 1389\n",
      "150000 gPol1 2571\n",
      "200000 gPol2 3346\n",
      "250000 EnvT3 7343\n",
      "300000 EnvV4 7431\n",
      "350000 EnvV5 7614\n",
      "400000 gRNA Gag D 1061\n",
      "450000 CA3 1617\n",
      "500000 RT1 2861\n",
      "550000 gRNA-RRE 7782\n",
      "600000 gGag1 1389\n",
      "650000 gTatRev 5970\n",
      "700000 gP24-T2  1465\n",
      "750000 T10  8532\n",
      "800000 T5 2805\n",
      "850000 T6 3114\n",
      "900000 LTR-F 414\n",
      "950000 Gag-A 1465\n",
      "1000000 LTR-M 312\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Citation</th>\n",
       "      <th>Direc</th>\n",
       "      <th>HasPam</th>\n",
       "      <th>Hit</th>\n",
       "      <th>Name</th>\n",
       "      <th>Region</th>\n",
       "      <th>Score</th>\n",
       "      <th>SeqID</th>\n",
       "      <th>gRNA</th>\n",
       "      <th>gStart</th>\n",
       "      <th>gStop</th>\n",
       "      <th>ProtoPam</th>\n",
       "      <th>HasN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>26607397</td>\n",
       "      <td>Forward</td>\n",
       "      <td>True</td>\n",
       "      <td>GACAAGATATCCTTGATCTGTGG</td>\n",
       "      <td>LTR-1</td>\n",
       "      <td>LTR</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>A0009-R00-PBMC-Genomic-LTR</td>\n",
       "      <td>GACAAGATATCCTTGATCTG.NGG</td>\n",
       "      <td>28</td>\n",
       "      <td>47</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26607397</td>\n",
       "      <td>Forward</td>\n",
       "      <td>True</td>\n",
       "      <td>GAAAAGAGATCCTTGATCTGTGG</td>\n",
       "      <td>LTR-1</td>\n",
       "      <td>LTR</td>\n",
       "      <td>0.98600</td>\n",
       "      <td>A0026-R06-PBMC-Genomic-LTR</td>\n",
       "      <td>GACAAGATATCCTTGATCTG.NGG</td>\n",
       "      <td>28</td>\n",
       "      <td>47</td>\n",
       "      <td>0.98600</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>26607397</td>\n",
       "      <td>Forward</td>\n",
       "      <td>True</td>\n",
       "      <td>GACAAGATATCCTTGATCTGTGG</td>\n",
       "      <td>LTR-1</td>\n",
       "      <td>LTR</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>A0041-R05-PBMC-Genomic-LTR</td>\n",
       "      <td>GACAAGATATCCTTGATCTG.NGG</td>\n",
       "      <td>28</td>\n",
       "      <td>47</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>26607397</td>\n",
       "      <td>Forward</td>\n",
       "      <td>True</td>\n",
       "      <td>GACAAGACATCCTTGATCTGTGG</td>\n",
       "      <td>LTR-1</td>\n",
       "      <td>LTR</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>A0044-R06-PBMC-Genomic-LTR</td>\n",
       "      <td>GACAAGATATCCTTGATCTG.NGG</td>\n",
       "      <td>28</td>\n",
       "      <td>47</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>26607397</td>\n",
       "      <td>Forward</td>\n",
       "      <td>True</td>\n",
       "      <td>GACAAGACATCCTTGATNNGTGG</td>\n",
       "      <td>LTR-1</td>\n",
       "      <td>LTR</td>\n",
       "      <td>0.06174</td>\n",
       "      <td>A0044-R07-PBMC-Genomic-LTR</td>\n",
       "      <td>GACAAGATATCCTTGATCTG.NGG</td>\n",
       "      <td>28</td>\n",
       "      <td>47</td>\n",
       "      <td>0.06174</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Citation    Direc HasPam                      Hit   Name Region    Score  \\\n",
       "0  26607397  Forward   True  GACAAGATATCCTTGATCTGTGG  LTR-1    LTR  1.00000   \n",
       "1  26607397  Forward   True  GAAAAGAGATCCTTGATCTGTGG  LTR-1    LTR  0.98600   \n",
       "2  26607397  Forward   True  GACAAGATATCCTTGATCTGTGG  LTR-1    LTR  1.00000   \n",
       "3  26607397  Forward   True  GACAAGACATCCTTGATCTGTGG  LTR-1    LTR  1.00000   \n",
       "4  26607397  Forward   True  GACAAGACATCCTTGATNNGTGG  LTR-1    LTR  0.06174   \n",
       "\n",
       "                        SeqID                      gRNA  gStart  gStop  \\\n",
       "0  A0009-R00-PBMC-Genomic-LTR  GACAAGATATCCTTGATCTG.NGG      28     47   \n",
       "1  A0026-R06-PBMC-Genomic-LTR  GACAAGATATCCTTGATCTG.NGG      28     47   \n",
       "2  A0041-R05-PBMC-Genomic-LTR  GACAAGATATCCTTGATCTG.NGG      28     47   \n",
       "3  A0044-R06-PBMC-Genomic-LTR  GACAAGATATCCTTGATCTG.NGG      28     47   \n",
       "4  A0044-R07-PBMC-Genomic-LTR  GACAAGATATCCTTGATCTG.NGG      28     47   \n",
       "\n",
       "   ProtoPam   HasN  \n",
       "0   1.00000  False  \n",
       "1   0.98600  False  \n",
       "2   1.00000  False  \n",
       "3   1.00000  False  \n",
       "4   0.06174   True  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "num = 0\n",
    "scores = []\n",
    "\n",
    "# Run through each gRNA\n",
    "for _, row in ex_df.iterrows():\n",
    "    \n",
    "    # Extract all sequences that overlap the target region\n",
    "    mask = (seq_df['sStart'] < (row['Start']-20)) & (seq_df['sStop'] > (row['Stop']+20))\n",
    "    \n",
    "    for _, seq_row in seq_df.ix[mask].iterrows():\n",
    "        num += 1\n",
    "        if num % 50000 == 0:\n",
    "            print num, row['Name'], row['Start']\n",
    "        \n",
    "        # Check this sequence for the presence of this gRNA\n",
    "        hit, score, direc, has_pam = check_whole_seq(seq_row['SeqR'].seq, row['gRNA'])\n",
    "    \n",
    "        # Extract some information\n",
    "        scores.append({'Hit': str(hit),\n",
    "                       'Score': score,\n",
    "                       'gRNA': row['gRNA'],\n",
    "                       'Citation': row['Citation'],\n",
    "                       'Direc': direc,\n",
    "                       'SeqID': seq_row['Name'],\n",
    "                       'gStart': row['Start'],\n",
    "                       'gStop': row['Stop'],\n",
    "                       'HasPam': has_pam,\n",
    "                       'Name': row['Name'],\n",
    "                       'Region': row['Region']})\n",
    "\n",
    "score_df = pd.DataFrame(scores)\n",
    "# Since the PAM is required, any sequence without a PAM will have its value zeroed out\n",
    "score_df['ProtoPam'] = score_df['Score']*(score_df['HasPam'].astype(float))\n",
    "\n",
    "# We need to exclude sequences which have Ns in the target region\n",
    "score_df['HasN'] = score_df['Hit'].str.contains('N')\n",
    "score_df.head()   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from scipy.stats import entropy\n",
    "\n",
    "def calc_entropy(ser):    \n",
    "    return entropy(ser.value_counts()/len(ser), base=2)\n",
    "\n",
    "\n",
    "# Group by each gRNA and calculate the entropy of the hits\n",
    "\n",
    "entropies = score_df.groupby('gRNA')['Hit'].agg(calc_entropy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Region</th>\n",
       "      <th>Citation</th>\n",
       "      <th>Name</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>gRNA</th>\n",
       "      <th>ProtoMean</th>\n",
       "      <th>ProtoBind</th>\n",
       "      <th>ProtoCut</th>\n",
       "      <th>HasPam</th>\n",
       "      <th>ProtoPamMean</th>\n",
       "      <th>ProtoPamBind</th>\n",
       "      <th>ProtoPamCut</th>\n",
       "      <th>Entropy</th>\n",
       "      <th>NumSeqs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LTR</td>\n",
       "      <td>23974631</td>\n",
       "      <td>T5</td>\n",
       "      <td>465</td>\n",
       "      <td>484</td>\n",
       "      <td>GTTAGACCAGATCTGAGCCT.NGG</td>\n",
       "      <td>0.776270</td>\n",
       "      <td>0.696307</td>\n",
       "      <td>0.688018</td>\n",
       "      <td>0.963828</td>\n",
       "      <td>0.769499</td>\n",
       "      <td>0.691032</td>\n",
       "      <td>0.682743</td>\n",
       "      <td>2.615312</td>\n",
       "      <td>1327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LTR</td>\n",
       "      <td>25049410</td>\n",
       "      <td>LTR-A</td>\n",
       "      <td>98</td>\n",
       "      <td>127</td>\n",
       "      <td>AGGGCCAGGGATCAGATATCCACTGACCTT.NGG</td>\n",
       "      <td>0.682982</td>\n",
       "      <td>0.628352</td>\n",
       "      <td>0.554278</td>\n",
       "      <td>0.947637</td>\n",
       "      <td>0.676152</td>\n",
       "      <td>0.623244</td>\n",
       "      <td>0.550447</td>\n",
       "      <td>5.249049</td>\n",
       "      <td>783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LTR</td>\n",
       "      <td>25049410</td>\n",
       "      <td>LTR-B</td>\n",
       "      <td>312</td>\n",
       "      <td>341</td>\n",
       "      <td>CCN.GAGTACTTCAAGAACTGCTGACATCGAGCT</td>\n",
       "      <td>0.431888</td>\n",
       "      <td>0.226766</td>\n",
       "      <td>0.214126</td>\n",
       "      <td>0.896654</td>\n",
       "      <td>0.422759</td>\n",
       "      <td>0.226022</td>\n",
       "      <td>0.214126</td>\n",
       "      <td>4.605056</td>\n",
       "      <td>1345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LTR</td>\n",
       "      <td>25049410</td>\n",
       "      <td>LTR-C</td>\n",
       "      <td>78</td>\n",
       "      <td>97</td>\n",
       "      <td>GATTGGCAGAACTACACACC.NGG</td>\n",
       "      <td>0.785196</td>\n",
       "      <td>0.754065</td>\n",
       "      <td>0.705285</td>\n",
       "      <td>0.983740</td>\n",
       "      <td>0.782252</td>\n",
       "      <td>0.752033</td>\n",
       "      <td>0.703252</td>\n",
       "      <td>2.660178</td>\n",
       "      <td>492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LTR</td>\n",
       "      <td>25049410</td>\n",
       "      <td>LTR-D</td>\n",
       "      <td>380</td>\n",
       "      <td>399</td>\n",
       "      <td>GCGTGGCCTGGGCGGGACTG.NGG</td>\n",
       "      <td>0.758716</td>\n",
       "      <td>0.754996</td>\n",
       "      <td>0.606218</td>\n",
       "      <td>0.975574</td>\n",
       "      <td>0.747624</td>\n",
       "      <td>0.745374</td>\n",
       "      <td>0.596595</td>\n",
       "      <td>4.206223</td>\n",
       "      <td>1351</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Region  Citation   Name  Start  Stop                                gRNA  \\\n",
       "0    LTR  23974631     T5    465   484            GTTAGACCAGATCTGAGCCT.NGG   \n",
       "1    LTR  25049410  LTR-A     98   127  AGGGCCAGGGATCAGATATCCACTGACCTT.NGG   \n",
       "2    LTR  25049410  LTR-B    312   341  CCN.GAGTACTTCAAGAACTGCTGACATCGAGCT   \n",
       "3    LTR  25049410  LTR-C     78    97            GATTGGCAGAACTACACACC.NGG   \n",
       "4    LTR  25049410  LTR-D    380   399            GCGTGGCCTGGGCGGGACTG.NGG   \n",
       "\n",
       "   ProtoMean  ProtoBind  ProtoCut    HasPam  ProtoPamMean  ProtoPamBind  \\\n",
       "0   0.776270   0.696307  0.688018  0.963828      0.769499      0.691032   \n",
       "1   0.682982   0.628352  0.554278  0.947637      0.676152      0.623244   \n",
       "2   0.431888   0.226766  0.214126  0.896654      0.422759      0.226022   \n",
       "3   0.785196   0.754065  0.705285  0.983740      0.782252      0.752033   \n",
       "4   0.758716   0.754996  0.606218  0.975574      0.747624      0.745374   \n",
       "\n",
       "   ProtoPamCut   Entropy  NumSeqs  \n",
       "0     0.682743  2.615312     1327  \n",
       "1     0.550447  5.249049      783  \n",
       "2     0.214126  4.605056     1345  \n",
       "3     0.703252  2.660178      492  \n",
       "4     0.596595  4.206223     1351  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "sum_data = []\n",
    "# Group by gRNA and Cititation and calculate aggregate stats\n",
    "for (grna, cit), rows in score_df.query('HasN == False').groupby(['gRNA', 'Citation']):\n",
    "    \n",
    "    rows = rows.groupby('SeqID', as_index=False).first()\n",
    "    \n",
    "    has_pam = rows['HasPam'].mean() # Fraction of hits with adjacent PAMs\n",
    "    num_seqs = len(rows) \n",
    "    proto_mean = rows['Score'].mean() # Average MIT score\n",
    "    proto_bind = (rows['Score']>0.5).mean() # Fraction of sequences with MIT > 0.5\n",
    "    proto_cut = (rows['Score']>0.75).mean() # Fraction of sequences with MIT > 0.75\n",
    "    \n",
    "    proto_pam_mean = rows['ProtoPam'].mean() # Average MIT score with missing PAMs counting as Zero\n",
    "    proto_pam_bind = (rows['ProtoPam']>0.5).mean() # Fraction of sequences with MIT > 0.5 and missing PAMs counting as Zero\n",
    "    proto_pam_cut = (rows['ProtoPam']>0.75).mean() # Fraction of sequences with MIT > 0.75 and missing PAMs counting as Zero\n",
    "    \n",
    "    # Collect aggregate data\n",
    "    sum_data.append({'gRNA': grna,\n",
    "                     'Citation': cit,\n",
    "                     'HasPam': has_pam,\n",
    "                     'NumSeqs': num_seqs,\n",
    "                     'ProtoMean': proto_mean,\n",
    "                     'ProtoCut': proto_cut,\n",
    "                     'ProtoBind': proto_bind,\n",
    "                     'ProtoPamMean': proto_pam_mean,\n",
    "                     'ProtoPamBind': proto_pam_bind,\n",
    "                     'ProtoPamCut': proto_pam_cut,\n",
    "                     'Start': rows['gStart'].iloc[0],\n",
    "                     'Stop': rows['gStop'].iloc[0],\n",
    "                     'Name': rows['Name'].iloc[0],\n",
    "                     'Entropy': entropies[grna],\n",
    "                     'Region': rows['Region'].iloc[0]})\n",
    "    \n",
    "order = ['Region',u'Citation', u'Name', u'Start', u'Stop', u'gRNA',   \n",
    "         u'ProtoMean', 'ProtoBind', u'ProtoCut', \n",
    "         u'HasPam',  u'ProtoPamMean', 'ProtoPamBind', u'ProtoPamCut', 'Entropy', u'NumSeqs']\n",
    "sum_df = pd.DataFrame(sum_data)[order].sort_values(by = ['Citation', 'Name']).reset_index(drop=True)\n",
    "sum_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Output raw results to Excel sheet for later processing\n",
    "\n",
    "sum_df.to_excel('results/summary_res_all_gRNAs.xlsx', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "import os\n",
    "\n",
    "def pathify(citation, name, base_dir = 'results/FrequencyTables/'):\n",
    "    \n",
    "    fname = '%i_%s.xlsx' % (citation, name.strip())\n",
    "    return os.path.join(base_dir, fname.replace(' ', '-'))\n",
    "    \n",
    "# Extract the nucleotide frequencies for all gRNAs\n",
    "\n",
    "for (cit, name), df in score_df.groupby(['Citation', 'Name']):\n",
    "\n",
    "    counts = defaultdict(float)\n",
    "    tot = 0\n",
    "    for n in df.query('HasN == False')['Hit'].str.upper().values:\n",
    "        tot += 1\n",
    "        for p, l in enumerate(n):\n",
    "            counts[(p, l)] += 1\n",
    "\n",
    "    order = 'ACGT'\n",
    "    d = []\n",
    "    for num in range(23):\n",
    "        d.append([counts[(num, l)] for l in order])\n",
    "\n",
    "    res = pd.DataFrame(d, columns = list(order), index=range(1, 24)).T/tot\n",
    "    res.to_excel(pathify(cit, name))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
