{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sbn\n",
    "from Bio import SeqIO\n",
    "import SearchTools\n",
    "\n",
    "%matplotlib inline\n",
    "sbn.set(font_scale=1.5)\n",
    "sbn.set_style('white')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Import\n",
    "\n",
    "Sequence data is read in from the Los Alamos Database that is stored in the included `.tar.gz` file. These alignments were produced by the LANL tools and are aranged such that the HXB2 sequence is the last in the file. The alignment information is used to determine the HXB2 positions of each sequence. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing hiv-2500-3000.fixed.fst\n",
      "Importing hiv-1100-1950.fixed.fst\n",
      "Importing hiv-8300-8900.fixed.fst\n",
      "Importing hiv-7100-7500.fixed.fst\n",
      "Importing hiv-5200-5600.fixed.fst\n",
      "Importing hiv-1-700.fixed.fst\n",
      "Importing hiv-8500-8800.fixed.fst\n",
      "Importing hiv-6200-6900.fixed.fst\n",
      "Importing hiv-700-1150.fixed.fst\n",
      "Importing hiv-2000-2500.fixed.fst\n",
      "Importing hiv-5700-6100.fixed.fst\n",
      "Importing hiv-9086-9717.fixed.fst\n",
      "Importing hiv-4100-4900.fixed.fst\n",
      "Importing hiv-3000-3500.fixed.fst\n",
      "Importing hiv-7500-7900.fixed.fst\n",
      "Importing hiv-3200-3500.fixed.fst\n",
      "Total sequences imported 390290\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>SeqR</th>\n",
       "      <th>sStart</th>\n",
       "      <th>sStop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>B.US.1997.ARES2.AB078005</td>\n",
       "      <td>(A, C, A, T, A, A, T, T, G, -, -, -, -, -, G, ...</td>\n",
       "      <td>2500</td>\n",
       "      <td>3000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>B.US.1985.Ba_L.AB221005</td>\n",
       "      <td>(A, C, A, T, A, A, T, T, G, -, -, -, -, -, G, ...</td>\n",
       "      <td>2500</td>\n",
       "      <td>3000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B.US.1985.Ba_L.AB253432</td>\n",
       "      <td>(A, C, A, T, A, A, T, T, G, -, -, -, -, -, G, ...</td>\n",
       "      <td>2500</td>\n",
       "      <td>3000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>B.US.1991.US2.AB485638</td>\n",
       "      <td>(A, C, A, T, A, A, T, T, G, -, -, -, -, -, G, ...</td>\n",
       "      <td>2500</td>\n",
       "      <td>3000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B.US.1991.US2.AB485639</td>\n",
       "      <td>(A, C, A, T, A, A, T, T, G, -, -, -, -, -, G, ...</td>\n",
       "      <td>2500</td>\n",
       "      <td>3000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Name  \\\n",
       "0  B.US.1997.ARES2.AB078005   \n",
       "1   B.US.1985.Ba_L.AB221005   \n",
       "2   B.US.1985.Ba_L.AB253432   \n",
       "3    B.US.1991.US2.AB485638   \n",
       "4    B.US.1991.US2.AB485639   \n",
       "\n",
       "                                                SeqR  sStart  sStop  \n",
       "0  (A, C, A, T, A, A, T, T, G, -, -, -, -, -, G, ...    2500   3000  \n",
       "1  (A, C, A, T, A, A, T, T, G, -, -, -, -, -, G, ...    2500   3000  \n",
       "2  (A, C, A, T, A, A, T, T, G, -, -, -, -, -, G, ...    2500   3000  \n",
       "3  (A, C, A, T, A, A, T, T, G, -, -, -, -, -, G, ...    2500   3000  \n",
       "4  (A, C, A, T, A, A, T, T, G, -, -, -, -, -, G, ...    2500   3000  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import tarfile\n",
    "from collections import Counter\n",
    "from copy import deepcopy\n",
    "from itertools import compress\n",
    "from collections import deque\n",
    "\n",
    "def get_start_stop(seqR, hxb2_pos):\n",
    "    \"\"\"Determine the HXB2 start-stop of the sequence using affine-gaps.\"\"\"\n",
    "    \n",
    "    it = zip(str(seqR.seq), hxb2_pos)\n",
    "    poses = compress(it, (x[0] != '-' for x in it))\n",
    "    start = poses.next()[1]\n",
    "    stop = deque(poses, maxlen=1)[0][1]\n",
    "    return start, stop\n",
    "\n",
    "\n",
    "seq_locs = []\n",
    "\n",
    "with tarfile.open('data/LANLdata.tar.gz', mode='r:gz') as tr:\n",
    "    for f in tr:\n",
    "        \n",
    "        parts = f.name.split('/')[-1].split('.')[0].split('-')\n",
    "        offset = int(parts[1])\n",
    "        print 'Importing', f.name\n",
    "        \n",
    "        seqs = list(SeqIO.parse(tr.extractfile(f), 'fasta'))\n",
    "        hxb2 = seqs[-1]\n",
    "\n",
    "        hxb2_pos = []\n",
    "        pos = offset-1\n",
    "        for hx_l in str(hxb2.seq):\n",
    "            pos += hx_l != '-'\n",
    "            hxb2_pos.append(pos)\n",
    "\n",
    "        for seqR in seqs[:-1]:\n",
    "            if len(seqR.seq.ungap('-')) > 50:\n",
    "                start, stop = get_start_stop(seqR, hxb2_pos)\n",
    "                seq_locs.append({'sStart': start,\n",
    "                                 'sStop': stop,\n",
    "                                 'SeqR': deepcopy(seqR),\n",
    "                                 'Name': seqR.id})\n",
    "\n",
    "            \n",
    "seq_df = pd.DataFrame(seq_locs)\n",
    "print 'Total sequences imported', len(seq_locs)\n",
    "seq_df.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    390290.000000\n",
       "mean        391.866123\n",
       "std         166.220899\n",
       "min           0.000000\n",
       "25%         261.000000\n",
       "50%         399.000000\n",
       "75%         500.000000\n",
       "max         850.000000\n",
       "dtype: float64"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(seq_df['sStart']-seq_df['sStop']).abs().describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Citation</th>\n",
       "      <th>Name</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>gRNA</th>\n",
       "      <th>Region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>23974631</td>\n",
       "      <td>T5</td>\n",
       "      <td>465</td>\n",
       "      <td>484</td>\n",
       "      <td>GTTAGACCAGATCTGAGCCT.NGG</td>\n",
       "      <td>LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>23974631</td>\n",
       "      <td>T6</td>\n",
       "      <td>343</td>\n",
       "      <td>362</td>\n",
       "      <td>GCTACAAGGGACTTTCCGCT.NGG</td>\n",
       "      <td>LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>25049410</td>\n",
       "      <td>LTR-A</td>\n",
       "      <td>98</td>\n",
       "      <td>127</td>\n",
       "      <td>AGGGCCAGGGATCAGATATCCACTGACCTT.NGG</td>\n",
       "      <td>LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>25049410</td>\n",
       "      <td>LTR-B</td>\n",
       "      <td>312</td>\n",
       "      <td>341</td>\n",
       "      <td>CCN.GAGTACTTCAAGAACTGCTGACATCGAGCT</td>\n",
       "      <td>LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>25049410</td>\n",
       "      <td>LTR-C</td>\n",
       "      <td>78</td>\n",
       "      <td>97</td>\n",
       "      <td>GATTGGCAGAACTACACACC.NGG</td>\n",
       "      <td>LTR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Citation   Name  Start  Stop                                gRNA Region\n",
       "0  23974631     T5    465   484            GTTAGACCAGATCTGAGCCT.NGG    LTR\n",
       "1  23974631     T6    343   362            GCTACAAGGGACTTTCCGCT.NGG    LTR\n",
       "2  25049410  LTR-A     98   127  AGGGCCAGGGATCAGATATCCACTGACCTT.NGG    LTR\n",
       "3  25049410  LTR-B    312   341  CCN.GAGTACTTCAAGAACTGCTGACATCGAGCT    LTR\n",
       "4  25049410  LTR-C     78    97            GATTGGCAGAACTACACACC.NGG    LTR"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ex_df = pd.read_excel('data/gRNAList.xlsx')\n",
    "ex_df['gRNA'] = ex_df['gRNA'].str.strip()\n",
    "ex_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sequence Processing\n",
    "\n",
    "Now that the gRNAs and sequences have been loaded we'll compare them. Each gRNA will be parsed to extract the protospacer region. For each gRNA we extract all sequences that overlap the target region. Then, the MIT penalty matrix is applied exauhstively across the entire sequence to determine the ideal binding location."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from Bio.Seq import reverse_complement\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_grna(gRNA):\n",
    "    \"\"\"Extract and normalize the length of the protospacer\"\"\"\n",
    "    \n",
    "    parts = gRNA.split('.')\n",
    "    if len(parts[1]) == 3:\n",
    "        direc = 'Forward'\n",
    "        proto = parts[0][-20:]\n",
    "    else:\n",
    "        direc = 'Reverse'\n",
    "        proto = parts[1][:20]\n",
    "        \n",
    "    if len(proto) < 20:\n",
    "        proto = 'N'*(20-len(proto)) + proto\n",
    "        \n",
    "    return direc, proto\n",
    "\n",
    "\n",
    "def check_whole_seq(seq, gRNA):\n",
    "    \"\"\"Apply the MIT penalty matrix across every postion on the sequence.\"\"\"\n",
    "    \n",
    "    direc, query = parse_grna(gRNA)\n",
    "    \n",
    "    # Normalize the orientation and ungap the sequence\n",
    "    if direc == 'Reverse':\n",
    "        nseq = seq.ungap('-')[::-1]\n",
    "        query = query[::-1]\n",
    "    else:\n",
    "        nseq = seq.ungap('-')\n",
    "    \n",
    "    # Check every 20-mer in the sequence.\n",
    "    scores = []\n",
    "    for start in range(len(nseq)-20):\n",
    "        scores.append(SearchTools.gRNA_score_hit(query, nseq[start:start+20]))\n",
    "\n",
    "    # Find the best score\n",
    "    scores = pd.Series(scores)\n",
    "    bscore, bloc = scores.max(), scores.idxmax()\n",
    "    \n",
    "    # Prep outputs\n",
    "    oseq = str(nseq[bloc:bloc+23])\n",
    "    has_pam = oseq.endswith('GG')\n",
    "    if direc == 'Reverse':\n",
    "        oseq = oseq.encode('ascii')[::-1]\n",
    "        has_pam = oseq.startswith('CC')\n",
    "    \n",
    "    return oseq, bscore, direc, has_pam\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50000 gEnv-T3 7343\n",
      "100000 gP24-T1  1254\n",
      "150000 gPol-T1 2374\n",
      "200000 gRev-T1 8378\n",
      "250000 gRev-T2 8568\n",
      "300000 T10  8532\n",
      "350000 T4  2255\n",
      "400000 T5 2805\n",
      "450000 T9 8406\n",
      "500000 sg362F.v4 355\n",
      "550000 gEnv1 7216\n",
      "600000 gGag2 1909\n",
      "650000 gPol1 2571\n",
      "700000 gPol2 3346\n",
      "750000 Gag-A 1465\n",
      "800000 LTR-S 213\n",
      "850000 EnvV2 6793\n",
      "900000 EnvV4 7431\n",
      "950000 T10 sgRNA 8513\n",
      "1000000 gRNA Gag D 1061\n",
      "1050000 CA3 1617\n",
      "1100000 RT1 2861\n",
      "1150000 gRNA-RRE 7782\n",
      "1200000 gGag1 1389\n",
      "1250000 gPol3 4149\n",
      "1300000 MA3  815\n",
      "1350000 PR1  2374\n",
      "1400000 PR2  2463\n",
      "1450000 PR3  2310\n",
      "1500000 PR4  2303\n",
      "1550000 PR5  2312\n",
      "1600000 RT2  2662\n",
      "1650000 RT3  2706\n",
      "1700000 RT4  2805\n",
      "1750000 LTR-1 83\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Citation</th>\n",
       "      <th>Direc</th>\n",
       "      <th>HasPam</th>\n",
       "      <th>Hit</th>\n",
       "      <th>Name</th>\n",
       "      <th>Region</th>\n",
       "      <th>Score</th>\n",
       "      <th>SeqID</th>\n",
       "      <th>gRNA</th>\n",
       "      <th>gStart</th>\n",
       "      <th>gStop</th>\n",
       "      <th>ProtoPam</th>\n",
       "      <th>HasN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>23974631</td>\n",
       "      <td>Forward</td>\n",
       "      <td>True</td>\n",
       "      <td>GTTAGACCAGATCTGAGCCTGGG</td>\n",
       "      <td>T5</td>\n",
       "      <td>LTR</td>\n",
       "      <td>1.000</td>\n",
       "      <td>A0001-R01-PBMC-Genomic-LTR</td>\n",
       "      <td>GTTAGACCAGATCTGAGCCT.NGG</td>\n",
       "      <td>465</td>\n",
       "      <td>484</td>\n",
       "      <td>1.000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>23974631</td>\n",
       "      <td>Forward</td>\n",
       "      <td>True</td>\n",
       "      <td>GTTAGACCAGATCTGAGCCTGGG</td>\n",
       "      <td>T5</td>\n",
       "      <td>LTR</td>\n",
       "      <td>1.000</td>\n",
       "      <td>A0001-R04-PBMC-Genomic-LTR</td>\n",
       "      <td>GTTAGACCAGATCTGAGCCT.NGG</td>\n",
       "      <td>465</td>\n",
       "      <td>484</td>\n",
       "      <td>1.000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>23974631</td>\n",
       "      <td>Forward</td>\n",
       "      <td>True</td>\n",
       "      <td>GTTAGACCAGATTTGAGCCTGGG</td>\n",
       "      <td>T5</td>\n",
       "      <td>LTR</td>\n",
       "      <td>0.387</td>\n",
       "      <td>A0002-R01-PBMC-Genomic-LTR</td>\n",
       "      <td>GTTAGACCAGATCTGAGCCT.NGG</td>\n",
       "      <td>465</td>\n",
       "      <td>484</td>\n",
       "      <td>0.387</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>23974631</td>\n",
       "      <td>Forward</td>\n",
       "      <td>True</td>\n",
       "      <td>GTTAGACCAGATTTGAGCCTGGG</td>\n",
       "      <td>T5</td>\n",
       "      <td>LTR</td>\n",
       "      <td>0.387</td>\n",
       "      <td>A0002-R03-PBMC-Genomic-LTR</td>\n",
       "      <td>GTTAGACCAGATCTGAGCCT.NGG</td>\n",
       "      <td>465</td>\n",
       "      <td>484</td>\n",
       "      <td>0.387</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>23974631</td>\n",
       "      <td>Forward</td>\n",
       "      <td>True</td>\n",
       "      <td>GTTAGACCAGATCTGAGCCTGGG</td>\n",
       "      <td>T5</td>\n",
       "      <td>LTR</td>\n",
       "      <td>1.000</td>\n",
       "      <td>A0002-R04-PBMC-Genomic-LTR</td>\n",
       "      <td>GTTAGACCAGATCTGAGCCT.NGG</td>\n",
       "      <td>465</td>\n",
       "      <td>484</td>\n",
       "      <td>1.000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Citation    Direc HasPam                      Hit Name Region  Score  \\\n",
       "0  23974631  Forward   True  GTTAGACCAGATCTGAGCCTGGG   T5    LTR  1.000   \n",
       "1  23974631  Forward   True  GTTAGACCAGATCTGAGCCTGGG   T5    LTR  1.000   \n",
       "2  23974631  Forward   True  GTTAGACCAGATTTGAGCCTGGG   T5    LTR  0.387   \n",
       "3  23974631  Forward   True  GTTAGACCAGATTTGAGCCTGGG   T5    LTR  0.387   \n",
       "4  23974631  Forward   True  GTTAGACCAGATCTGAGCCTGGG   T5    LTR  1.000   \n",
       "\n",
       "                        SeqID                      gRNA  gStart  gStop  \\\n",
       "0  A0001-R01-PBMC-Genomic-LTR  GTTAGACCAGATCTGAGCCT.NGG     465    484   \n",
       "1  A0001-R04-PBMC-Genomic-LTR  GTTAGACCAGATCTGAGCCT.NGG     465    484   \n",
       "2  A0002-R01-PBMC-Genomic-LTR  GTTAGACCAGATCTGAGCCT.NGG     465    484   \n",
       "3  A0002-R03-PBMC-Genomic-LTR  GTTAGACCAGATCTGAGCCT.NGG     465    484   \n",
       "4  A0002-R04-PBMC-Genomic-LTR  GTTAGACCAGATCTGAGCCT.NGG     465    484   \n",
       "\n",
       "   ProtoPam   HasN  \n",
       "0     1.000  False  \n",
       "1     1.000  False  \n",
       "2     0.387  False  \n",
       "3     0.387  False  \n",
       "4     1.000  False  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "num = 0\n",
    "scores = []\n",
    "\n",
    "# Run through each gRNA\n",
    "for _, row in ex_df.iterrows():\n",
    "    \n",
    "    # Extract all sequences that overlap the target region\n",
    "    mask = (seq_df['sStart'] < (row['Start']-20)) & (seq_df['sStop'] > (row['Stop']+20))\n",
    "    \n",
    "    for _, seq_row in seq_df.ix[mask].iterrows():\n",
    "        num += 1\n",
    "        if num % 50000 == 0:\n",
    "            print num, row['Name'], row['Start']\n",
    "        \n",
    "        # Check this sequence for the presence of this gRNA\n",
    "        hit, score, direc, has_pam = check_whole_seq(seq_row['SeqR'].seq, row['gRNA'])\n",
    "    \n",
    "        # Extract some information\n",
    "        scores.append({'Hit': str(hit),\n",
    "                       'Score': score,\n",
    "                       'gRNA': row['gRNA'],\n",
    "                       'Citation': row['Citation'],\n",
    "                       'Direc': direc,\n",
    "                       'SeqID': seq_row['Name'],\n",
    "                       'gStart': row['Start'],\n",
    "                       'gStop': row['Stop'],\n",
    "                       'HasPam': has_pam,\n",
    "                       'Name': row['Name'].strip(),\n",
    "                       'Region': row['Region']})\n",
    "\n",
    "score_df = pd.DataFrame(scores)\n",
    "# Since the PAM is required, any sequence without a PAM will have its value zeroed out\n",
    "score_df['ProtoPam'] = score_df['Score']*(score_df['HasPam'].astype(float))\n",
    "\n",
    "# We need to exclude sequences which have Ns in the target region\n",
    "score_df['HasN'] = score_df['Hit'].str.contains('N')\n",
    "score_df.head()   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from scipy.stats import entropy\n",
    "\n",
    "def calc_entropy(ser):    \n",
    "    return entropy(ser.value_counts()/len(ser), base=2)\n",
    "\n",
    "\n",
    "# Group by each gRNA and calculate the entropy of the hits\n",
    "\n",
    "entropies = score_df.groupby('gRNA')['Hit'].agg(calc_entropy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Region</th>\n",
       "      <th>Citation</th>\n",
       "      <th>Name</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>gRNA</th>\n",
       "      <th>Cutoff</th>\n",
       "      <th>ProtoMean</th>\n",
       "      <th>ProtoCut</th>\n",
       "      <th>HasPam</th>\n",
       "      <th>ProtoPamMean</th>\n",
       "      <th>ProtoPamCut</th>\n",
       "      <th>Entropy</th>\n",
       "      <th>NumSeqs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LTR</td>\n",
       "      <td>23974631</td>\n",
       "      <td>T5</td>\n",
       "      <td>465</td>\n",
       "      <td>484</td>\n",
       "      <td>GTTAGACCAGATCTGAGCCT.NGG</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.77627</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.963828</td>\n",
       "      <td>0.769499</td>\n",
       "      <td>0.963828</td>\n",
       "      <td>2.615312</td>\n",
       "      <td>1327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LTR</td>\n",
       "      <td>23974631</td>\n",
       "      <td>T5</td>\n",
       "      <td>465</td>\n",
       "      <td>484</td>\n",
       "      <td>GTTAGACCAGATCTGAGCCT.NGG</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.77627</td>\n",
       "      <td>0.978146</td>\n",
       "      <td>0.963828</td>\n",
       "      <td>0.769499</td>\n",
       "      <td>0.963075</td>\n",
       "      <td>2.615312</td>\n",
       "      <td>1327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LTR</td>\n",
       "      <td>23974631</td>\n",
       "      <td>T5</td>\n",
       "      <td>465</td>\n",
       "      <td>484</td>\n",
       "      <td>GTTAGACCAGATCTGAGCCT.NGG</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.77627</td>\n",
       "      <td>0.963828</td>\n",
       "      <td>0.963828</td>\n",
       "      <td>0.769499</td>\n",
       "      <td>0.951771</td>\n",
       "      <td>2.615312</td>\n",
       "      <td>1327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LTR</td>\n",
       "      <td>23974631</td>\n",
       "      <td>T5</td>\n",
       "      <td>465</td>\n",
       "      <td>484</td>\n",
       "      <td>GTTAGACCAGATCTGAGCCT.NGG</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.77627</td>\n",
       "      <td>0.876413</td>\n",
       "      <td>0.963828</td>\n",
       "      <td>0.769499</td>\n",
       "      <td>0.870384</td>\n",
       "      <td>2.615312</td>\n",
       "      <td>1327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LTR</td>\n",
       "      <td>23974631</td>\n",
       "      <td>T5</td>\n",
       "      <td>465</td>\n",
       "      <td>484</td>\n",
       "      <td>GTTAGACCAGATCTGAGCCT.NGG</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.77627</td>\n",
       "      <td>0.874906</td>\n",
       "      <td>0.963828</td>\n",
       "      <td>0.769499</td>\n",
       "      <td>0.868877</td>\n",
       "      <td>2.615312</td>\n",
       "      <td>1327</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Region  Citation Name  Start  Stop                      gRNA  Cutoff  \\\n",
       "0    LTR  23974631   T5    465   484  GTTAGACCAGATCTGAGCCT.NGG    0.00   \n",
       "1    LTR  23974631   T5    465   484  GTTAGACCAGATCTGAGCCT.NGG    0.05   \n",
       "2    LTR  23974631   T5    465   484  GTTAGACCAGATCTGAGCCT.NGG    0.10   \n",
       "3    LTR  23974631   T5    465   484  GTTAGACCAGATCTGAGCCT.NGG    0.15   \n",
       "4    LTR  23974631   T5    465   484  GTTAGACCAGATCTGAGCCT.NGG    0.20   \n",
       "\n",
       "   ProtoMean  ProtoCut    HasPam  ProtoPamMean  ProtoPamCut   Entropy  NumSeqs  \n",
       "0    0.77627  1.000000  0.963828      0.769499     0.963828  2.615312     1327  \n",
       "1    0.77627  0.978146  0.963828      0.769499     0.963075  2.615312     1327  \n",
       "2    0.77627  0.963828  0.963828      0.769499     0.951771  2.615312     1327  \n",
       "3    0.77627  0.876413  0.963828      0.769499     0.870384  2.615312     1327  \n",
       "4    0.77627  0.874906  0.963828      0.769499     0.868877  2.615312     1327  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "cuts = np.arange(0, 1.05, 0.05)\n",
    "\n",
    "sum_data = []\n",
    "# Group by gRNA and Cititation and calculate aggregate stats\n",
    "for (grna, cit), rows in score_df.query('HasN == False').groupby(['gRNA', 'Citation']):\n",
    "    \n",
    "    rows = rows.groupby('SeqID', as_index=False).first()\n",
    "    \n",
    "    has_pam = rows['HasPam'].mean() # Fraction of hits with adjacent PAMs\n",
    "    num_seqs = len(rows)\n",
    "    \n",
    "    for cut in cuts:\n",
    "        proto_mean = rows['Score'].mean() # Average MIT score\n",
    "        proto_cut = (rows['Score']>cut).mean() # Fraction of sequences with MIT > cutoff\n",
    "\n",
    "        proto_pam_mean = rows['ProtoPam'].mean() # Average MIT score with missing PAMs counting as Zero\n",
    "        proto_pam_cut = (rows['ProtoPam']>cut).mean() # Fraction of sequences with MIT > cut and missing PAMs counting as Zero\n",
    "\n",
    "        # Collect aggregate data\n",
    "        sum_data.append({'gRNA': grna,\n",
    "                         'Cutoff': cut,\n",
    "                         'Citation': cit,\n",
    "                         'HasPam': has_pam,\n",
    "                         'NumSeqs': num_seqs,\n",
    "                         'ProtoMean': proto_mean,\n",
    "                         'ProtoCut': proto_cut,\n",
    "                         'ProtoPamMean': proto_pam_mean,\n",
    "                         'ProtoPamCut': proto_pam_cut,\n",
    "                         'Start': rows['gStart'].iloc[0],\n",
    "                         'Stop': rows['gStop'].iloc[0],\n",
    "                         'Name': rows['Name'].iloc[0],\n",
    "                         'Entropy': entropies[grna],\n",
    "                         'Region': rows['Region'].iloc[0]})\n",
    "    \n",
    "order = ['Region', u'Citation', u'Name', u'Start', u'Stop', u'gRNA', 'Cutoff',\n",
    "         u'ProtoMean', 'ProtoCut', \n",
    "         u'HasPam',  u'ProtoPamMean', 'ProtoPamCut', 'Entropy', u'NumSeqs']\n",
    "sum_df = pd.DataFrame(sum_data)[order].sort_values(by = ['Citation', 'Name']).reset_index(drop=True)\n",
    "sum_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Output raw results to Excel sheet for later processing\n",
    "\n",
    "sum_df.to_excel('results/summary_res_all_gRNAs.xlsx', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "import os\n",
    "\n",
    "def pathify(citation, name, base_dir = 'results/FrequencyTables/'):\n",
    "    \n",
    "    fname = '%i_%s.xlsx' % (citation, name.strip())\n",
    "    return os.path.join(base_dir, fname.replace(' ', '-'))\n",
    "    \n",
    "# Extract the nucleotide frequencies for all gRNAs\n",
    "\n",
    "for (cit, name), df in score_df.groupby(['Citation', 'Name']):\n",
    "\n",
    "    counts = defaultdict(float)\n",
    "    tot = 0\n",
    "    for n in df.query('HasN == False')['Hit'].str.upper().values:\n",
    "        tot += 1\n",
    "        for p, l in enumerate(n):\n",
    "            counts[(p, l)] += 1\n",
    "\n",
    "    order = 'ACGT'\n",
    "    d = []\n",
    "    for num in range(23):\n",
    "        d.append([counts[(num, l)] for l in order])\n",
    "\n",
    "    res = pd.DataFrame(d, columns = list(order), index=range(1, 24)).T/tot\n",
    "    res.to_excel(pathify(cit, name))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
